### ISSUE TITLE
-D: does not work as in 0.8.0

### ISSUE THREAD

**Issue by foldl** [ziglang/zig#10430]
### Zig Version

0.9.0

### Steps to Reproduce

I have been using -D to do conditional compilation in 0.8.0 like this:

```zig
const enable_print = print_info == 1;
const print = printf_from_c;
fn foo(...) ?u16 {
    var bitmap: u16 = 0xffff;
    ... 
    if (enable_print) {
            print("bla bla bla: %04x\n", bitmap);
    }
    return bitmap;
}
```

I can compile the code using either `zig build-obj -Dprint_info=1 ...` or `zig build-obj -Dprint_info=0 ...` to turn on/off
the `print`s.

### Expected Behavior

I expect 0.9.0 should also work since I failed to find something about this in
 the [Changes](https://ziglang.org/download/0.9.0/release-notes.html) of 0.9.0.

### Actual Behavior

It complains that:

```
error: use of undeclared identifier 'print_info'
```

**Comment by Vexu** [comment]
`-D` has only ever worked for defining C macros (except in `zig build` where it is used to pass arguments to the build script), perhaps you were using `@cImport` in 0.8.0?

**Comment by SpexGuy** [comment]
If this code is never analyzed, the 0.8.0 compiler wouldn't have rejected it.  But if you were to actually call `foo` in 0.8.0 you would have seen a similar error.  `-D` has never worked in the way you are expecting.

**Comment by leecannon** [comment]
Is it possible you were using `usingnamespace` on a `@cImport`? Previously it would have brought c macros into scope.

**Comment by foldl** [comment]
I don't know why this issue is closed. What I reported is what has actually happened.

If someone thought this is impossible to happen, could you ask for a full test case?

Now, here is a full test case, even though no one asked for it, and this issue had been closed:

1. `main.c`

    ```c
    #include <stdio.h>
    #include <stdint.h>
    
    extern uint16_t foo(uint16_t v);
    
    void main(void)
    {
        printf("%d\n", foo(0xffff));
    }
    ```

1. `zbug.zig`

    ```zig
    const std = @import("std");
    
    pub usingnamespace @cImport({
        @cInclude("aaa.h");
    });
    
    const enable_print = print_info == 1;
    const print = printf;
    
    export fn foo(v: u16) u16 {
        var bitmap: u16 = 0xfaff;
        if (enable_print) {
            print("bla bla bla: %04x\n", v);
        }
        return bitmap & v;
    }
    
    pub fn panic(message: []const u8, stack_trace: ?*std.builtin.StackTrace) noreturn {
        while (true) {}
    }
    ```
1. `aaa.h`

    ```c
    void printf(const char *fmt, ...);
    ```

1. test commands

    ```shell
    zig build-obj -Dprint_info=0 zbug.zig -I . -femit-bin=zbug.o
    gcc main.c zbug.o
    ```

    Using 0.8.0, toggle `print_info` between 0 and 1, and recompile, we can see that it works.

**Comment by SpexGuy** [comment]
Ah, this is the effect of an intentional breaking change to the meaning of `usingnamespace`.  The fixed version for 0.9.0 looks something like this:
```zig
const c = @cImport({
    @cInclude("aaa.h");
});
const printf = c.printf;
const print_info = c.print_info;
```

The `-D` option adds a define to C files and headers.  translate-c translates defines, since they are part of the exposed API of the C module.  So the `-D` parameter was "passed through" the C header and back into the zig namespace through the `usingnamespace` declaration.  However, `usingnamespace` no longer adds unqualified names to the current namespace in 0.9.0.  Instead it allows qualified references through the parent struct to find them.  So `@This().print_info` would work with the old code but `print_info` with no qualifier will not anymore.

**Comment by foldl** [comment]
OK. Now it seems reasonable to me.
